File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
//Given an Array, find the first duplicate value that occurs. If there are no duplicates, return -1.
//Write a function that takes in one array as an argument, find the first occurance of a duplicate and return it.
//If there is not a duplicate in the array, return -1

function findFirstDuplicate(arr) {
  let arrSet = new Set()
  for (let i= 0; i < arr.length; i++) {
    if (arrSet.has(arr[i])) return arr[i]
     arrSet.add(arr[i])
  }
  return -1
}



if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([]));

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate(["a", "b", "c", "d"]));

  console.log("Expecting: 7");
  console.log("=>", findFirstDuplicate([7, 7, 7]));

// provided test cases
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
//initialize an empty Set object as a place to store unique values as I iterate 
//iterate through the array/input 
//check on every loop through the array add it to the set if it is not there yet
//if it is there, I found the dublicate and return it
//if I get to the end of the array and have not found a duplicate, return -1

// And a written explanation of your solution
//The Set constructor allows for creating Set objects that store unique values.
//additionally it gives me access to two other methods set.has() set.add()
//I can initialize an empty set object and then iterate through my array/input to check for duplicates and check or add to my set object
// therefore I am creating my set object as I go and checking for duplicates
//if the the current element of my array I am looking at on one pass of my loop is in my Set already .has() 
//I have found my first duplicate, if it is not there yet, I add it and keep looking through unitl I either
//find an element that is already thee (my duplicate) or if I reach the end of my input arry and have not found a duplicate
//I can return whatever I want, in this case -1

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  hash_count = Hash.new(0)

  arr.each do |e|
    hash_count[e] = hash_count[e] += 1
    return e if hash_count[e] > 1
  end
  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  
  puts

  puts "Expecting: -1"
  puts find_first_duplicate([])

  puts

  puts "Expecting: -1"
  puts find_first_duplicate([7])


  # Don't forget to add your own!
end
#given an array, find the furst duplicate that occurs and return it.

# Please add your pseudocode to this file
# create a counter hash that keeps track of values
# pass in default of 0 so can add on to the hash and increment a counter
# iterate over the array and call counter hash with each element passed in as key
# and increment as soon as find a duplicate
# once there is a value in hash, return that key (dupliucate element)
# else return -1



# And a written explanation of your solution
# iterate through the arr passed in as an argumnet passing in each element as a key and value of 1
# when the incrementer hits a repeated value hash[e] = hash[e], it will increment and then give a value greater than one. Can then return the value 


